tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -40,238 +86,263 @@ auto read_json(std::filesystem::path const& path) {
+template <typename T>
+bool compare_value(T const& ref_attribute_value, T const& attribute_value, double atol, double rtol) {
+    auto is_within_tolerance = [atol, rtol](std::complex<double> element, std::complex<double> ref_element) {
+        return std::abs(element - ref_element) < std::abs(ref_element) * rtol + atol;
+    };
 
-    switch (attr.ctype) {
-    case c_int32:
-        return get_as_string<int32_t>(raw_data_ptr, attr, obj);
-    case c_int8:
-        return get_as_string<int8_t>(raw_data_ptr, attr, obj);
-    case c_double:
-        return get_as_string<double>(raw_data_ptr, attr, obj);
-    case c_double3:
-        return get_as_string<RealValue<asymmetric_t>>(raw_data_ptr, attr, obj);
-    default:
-        return "<unknown value type>"s;
+    if constexpr (std::is_same_v<std::decay_t<T>, std::array<double, 3>>) {
+        return std::ranges::equal(attribute_value, ref_attribute_value, is_within_tolerance);
+    } else if constexpr (std::is_same_v<std::decay_t<T>, double>) {
+        return is_within_tolerance(attribute_value, ref_attribute_value);
+    } else {
+        return ref_attribute_value == attribute_value;
     }
 }
 
-template <symmetry_tag sym>
-bool check_angle_and_magnitude(RawDataConstPtr reference_result_ptr, RawDataConstPtr result_ptr,
-                               MetaAttribute const& angle_attr, MetaAttribute const& mag_attr, double atol, double rtol,
-                               Idx obj) {
-    RealValue<sym> mag{};
-    RealValue<sym> mag_ref{};
-    RealValue<sym> angle{};
-    RealValue<sym> angle_ref{};
-    mag_attr.get_value(result_ptr, &mag, obj);
-    mag_attr.get_value(reference_result_ptr, &mag_ref, obj);
-    angle_attr.get_value(result_ptr, &angle, obj);
-    angle_attr.get_value(reference_result_ptr, &angle_ref, obj);
-    ComplexValue<sym> const result = mag * exp(1.0i * angle);
-    ComplexValue<sym> const result_ref = mag_ref * exp(1.0i * angle_ref);
-    if constexpr (is_symmetric_v<sym>) {
-        return cabs(result - result_ref) < (cabs(result_ref) * rtol + atol);
+template <typename T>
+void check_results(T const& ref_attribute_value, T const& attribute_value, T const& ref_possible_attribute_value,
+                   T const& possible_attribute_value, bool const& is_angle, Idx scenario_idx, Idx obj,
+                   std::string const& component_name, std::string const& attribute_name, double const& dynamic_atol,
+                   double const& rtol) {
+    bool const match =
+        is_angle
+            ? check_angle_and_magnitude<decltype(ref_attribute_value)>(ref_attribute_value, attribute_value,
+                                                                       ref_possible_attribute_value,
+                                                                       possible_attribute_value, dynamic_atol, rtol)
+            : compare_value<decltype(ref_attribute_value)>(ref_attribute_value, attribute_value, dynamic_atol, rtol);
+    if (match) {
+        CHECK(match);
```

Comment:
This line is redundant

Reply:
In the context of did this test fail or not, yes, but it is very useful for debugging, so I would advocate to leave it in.

Reply:
I am not sure that's a convincing argument. You could also simply do:
```suggestion
    if (!match) {
        //...
```

Reply:
After thinking again about this, I think you are right, this is just redundant. Change will be reflected in following commit.